### PR TITLE
Optimize SEL search efficiency

### PR DESCRIPTION
**Pull request by zhongwuzw**
## Changes in this pull request

Optimize `SEL` search efficiency, reduced the time complexity from O (n) to O (1).

Not need tests.

### Checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I added tests, an experiment, or detailed why my change isn't tested.
- [x] I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.
- [x] I have reviewed the [contributing guide](https://github.com/Instagram/IGListKit/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

File: Source/Internal/IGListAdapterProxy.m [hunk 1]
```diff
@@ -12,30 +12,34 @@
 #import <IGListKit/IGListAssert.h>
 
 /**
- Define messages that you want the IGListAdapter object to intercept. Pattern copied from
- https://github.com/facebook/AsyncDisplayKit/blob/7b112a2dcd0391ddf3671f9dcb63521f554b78bd/AsyncDisplayKit/ASCollectionView.mm#L34-L53
+ Define messages that you want the IGListAdapter object to intercept.
```
Comment (jessesquires) [inline Source/Internal/IGListAdapterProxy.m]: let's keep the original comment.

File: Source/Internal/IGListAdapterProxy.m [hunk 2]
```diff
@@ -12,30 +12,34 @@
...
-            sel == @selector(collectionView:layout:minimumLineSpacingForSectionAtIndex:) ||
-            sel == @selector(collectionView:layout:referenceSizeForFooterInSection:) ||
-            sel == @selector(collectionView:layout:referenceSizeForHeaderInSection:) ||
-            // UIScrollViewDelegate
-            sel == @selector(scrollViewDidScroll:) ||
-            sel == @selector(scrollViewWillBeginDragging:) ||
-            sel == @selector(scrollViewDidEndDragging:willDecelerate:) ||
-            sel == @selector(scrollViewDidEndDecelerating:)
-            );
+    static NSSet<NSString *> *sels;
```
Comment (jessesquires) [inline Source/Internal/IGListAdapterProxy.m]: let's spell-out `selectors`

**Comment by rnystrom** [comment]
This makes sense to me, though I'm curious @zhongwuzw did you do any instrumentation on the efficiency of this? Just curious what the cost of the linear pointer equality checks vs acquiring a lock + lookup is against each other?

I ask b/c even though this is straightforward, I'm skeptical about making changes w/out evidence that it's actually better.

**Comment by zhongwuzw** [comment]
@rnystrom , 👍 , honestly, I did not test it before, just based on my theoretical knowledge. After your comment, I tested like below, 

**The output:**

1. Compare using pointer check:
0.210490
0.226435
0.232305
0.278115

2. Compare using `NSSet`:
1.165643
0.947413
1.003125
0.999605

**Conclusion:**
`NSSet` seems not good vs pointer compare. So maybe we should still use pointer to check. But I want to update the order of these three protocol, put `UIScrollViewDelegate` methods to front, because the times of call are more than others. 😄 


```
static BOOL isInterceptedSelector(SEL sel) {
    static NSSet<NSString *> *selectors;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        selectors = [NSSet setWithObjects:
                     // UICollectionViewDelegate
                     NSStringFromSelector(@selector(collectionView:didSelectItemAtIndexPath:)),
                     NSStringFromSelector(@selector(collectionView:willDisplayCell:forItemAtIndexPath:)),
                     NSStringFromSelector(@selector(collectionView:didEndDisplayingCell:forItemAtIndexPath:)),
                     NSStringFromSelector(@selector(collectionView:didHighlightItemAtIndexPath:)),
                     NSStringFromSelector(@selector(collectionView:didUnhighlightItemAtIndexPath:)),
                     // UICollectionViewDelegateFlowLayout
                     NSStringFromSelector(@selector(collectionView:layout:sizeForItemAtIndexPath:)),
                     NSStringFromSelector(@selector(collectionView:layout:insetForSectionAtIndex:)),
                     NSStringFromSelector(@selector(collectionView:layout:minimumInteritemSpacingForSectionAtIndex:)),
                     NSStringFromSelector(@selector(collectionView:layout:minimumLineSpacingForSectionAtIndex:)),
                     NSStringFromSelector(@selector(collectionView:layout:referenceSizeForFooterInSection:)),
                     NSStringFromSelector(@selector(collectionView:layout:referenceSizeForHeaderInSection:)),
                     // UIScrollViewDelegate
                     NSStringFromSelector(@selector(scrollViewDidScroll:)),
                     NSStringFromSelector(@selector(scrollViewWillBeginDragging:)),
                     NSStringFromSelector(@selector(scrollViewDidEndDragging:willDecelerate:)),
                     NSStringFromSelector(@selector(scrollViewDidEndDecelerating:)),
                     nil];
    });
    return [selectors containsObject:NSStringFromSelector(sel)];
}

static BOOL isInterceptedSelector1(SEL sel) {
    return (
            // UICollectionViewDelegate
            sel == @selector(collectionView:didSelectItemAtIndexPath:) ||
            sel == @selector(collectionView:willDisplayCell:forItemAtIndexPath:) ||
            sel == @selector(collectionView:didEndDisplayingCell:forItemAtIndexPath:) ||
            sel == @selector(collectionView:didHighlightItemAtIndexPath:) ||
            sel == @selector(collectionView:didUnhighlightItemAtIndexPath:) ||
            // UICollectionViewDelegateFlowLayout
            sel == @selector(collectionView:layout:sizeForItemAtIndexPath:) ||
            sel == @selector(collectionView:layout:insetForSectionAtIndex:) ||
            sel == @selector(collectionView:layout:minimumInteritemSpacingForSectionAtIndex:) ||
            sel == @selector(collectionView:layout:minimumLineSpacingForSectionAtIndex:) ||
            sel == @selector(collectionView:layout:referenceSizeForFooterInSection:) ||
            sel == @selector(collectionView:layout:referenceSizeForHeaderInSection:) ||
            // UIScrollViewDelegate
            sel == @selector(scrollViewDidScroll:) ||
            sel == @selector(scrollViewWillBeginDragging:) ||
            sel == @selector(scrollViewDidEndDragging:willDecelerate:) ||
            sel == @selector(scrollViewDidEndDecelerating:)
            );
}

- (void)benchmark{
    NSSet *selectors;
    selectors = [NSSet setWithObjects:
                 // UICollectionViewDelegate
                 NSStringFromSelector(@selector(collectionView:didSelectItemAtIndexPath:)),
                 NSStringFromSelector(@selector(collectionView:willDisplayCell:forItemAtIndexPath:)),
                 NSStringFromSelector(@selector(collectionView:didEndDisplayingCell:forItemAtIndexPath:)),
                 NSStringFromSelector(@selector(collectionView:didHighlightItemAtIndexPath:)),
                 NSStringFromSelector(@selector(collectionView:didUnhighlightItemAtIndexPath:)),
                 // UICollectionViewDelegateFlowLayout
                 NSStringFromSelector(@selector(collectionView:layout:sizeForItemAtIndexPath:)),
                 NSStringFromSelector(@selector(collectionView:layout:insetForSectionAtIndex:)),
                 NSStringFromSelector(@selector(collectionView:layout:minimumInteritemSpacingForSectionAtIndex:)),
                 NSStringFromSelector(@selector(collectionView:layout:minimumLineSpacingForSectionAtIndex:)),
                 NSStringFromSelector(@selector(collectionView:layout:referenceSizeForFooterInSection:)),
                 NSStringFromSelector(@selector(collectionView:layout:referenceSizeForHeaderInSection:)),
                 // UIScrollViewDelegate
                 NSStringFromSelector(@selector(scrollViewDidScroll:)),
                 NSStringFromSelector(@selector(scrollViewWillBeginDragging:)),
                 NSStringFromSelector(@selector(scrollViewDidEndDragging:willDecelerate:)),
                 NSStringFromSelector(@selector(scrollViewDidEndDecelerating:)),
                 nil];
    
    NSTimeInterval startInterval = [[NSDate date] timeIntervalSince1970];
    
    for (int i = 0; i < 100000; i++) {
        for (NSString *selStr in selectors) {
            SEL sel = NSSelectorFromString(selStr);
            BOOL intercepted = isInterceptedSelector1(sel);
        }
    }
    
    NSTimeInterval endInterval = [[NSDate date] timeIntervalSince1970];
    NSLog(@"time consume is %f", endInterval - startInterval);

}
```

**Comment by jessesquires** [comment]
nice 😎

**Comment by zhongwuzw** [comment]
@rnystrom , I found the main performance overhead is `NSString` compare for `NSSet` approach, secondly is `Set`, cause of hash computation, dispatch_once maybe not too big effect.

**Pull request closed without merging**